d2/src/main/java/com/linkedin/d2/balancer/util/WarmUpLoadBalancer.java

```diff
@@ -242,8 +220,50 @@ void execute()
         return;
       }
 
-      LOG.info("{}/{} Starting to warm up service {}", new Object[]{_requestStartedCount.incrementAndGet(), _serviceNames.size(), serviceName});
+      DualReadModeProvider.DualReadMode dualReadMode = DualReadModeProvider.DualReadMode.OLD_LB_ONLY;
+
+      if (_dualReadStateManager != null)
+      {
+        DualReadModeProvider dualReadModeProvider = _dualReadStateManager.getDualReadModeProvider();
+        dualReadMode = dualReadModeProvider.getDualReadMode(serviceName);
+        _dualReadStateManager.updateService(serviceName, dualReadMode);
+
+        // dual read mode can be fine-tuned per service, check if the specific mode for this service should start
+        // the warm-up. If not, skip this service.
+        if (!isModeToWarmUp(dualReadMode, _isIndis))
+        {
+          return;
+        }
+      }
+
+      LOG.info("{}/{} Starting to warm up service {}", _requestStartedCount.incrementAndGet(), _serviceNames.size(),
+          serviceName);
 
+      // warm up dual read mode for the service and its belonging cluster. This is needed so that when the actual data
+      // of service/cluster/uri is received, they can be reported to dual read monitoring under dual read mode.
+      // NOTE: this will actually fetch the service data SYNCHRONOUSLY.
+      if (_dualReadStateManager != null)
+      {
+        DualReadModeProvider.DualReadMode finalDualReadMode = dualReadMode;
+        // NOTE: this call blocks!
+        getLoadBalancedServiceProperties(serviceName, new Callback<ServiceProperties>()
```

Comment:
Since this is blocking, this will break the current warmup flow in which executions finish instantaneously allowing multiple requests to be made simultaneously from the single-threaded executor.

Reply:
moved the whole fetching dual read mode logic to a separate executor task.